### PR TITLE
Whitelist sql column names

### PR DESCRIPTION
**Pull request by nickpell**
necessary to prevent sql injections

@czue @NoahCarnahan cc: @orangejenny

### CODE REVIEW

File: corehq/apps/userreports/reports/specs.py [hunk 1]
```diff
@@ -318,7 +318,7 @@ class FilterSpec(JsonObject):
     """
     type = StringProperty(required=True, choices=['date', 'numeric', 'choice_list', 'dynamic_choice_list'])
     slug = StringProperty(required=True, validators=_validate_filter_slug)  # this shows up as the ID in the filter HTML.
-    field = StringProperty(required=True)  # this is the actual column that is queried
+    field = StringProperty(required=True, validators=validate_sql_column_name)  # this is the actual column that is queried
```
Comment (dimagimon) [inline corehq/apps/userreports/reports/specs.py]: E501 line too long (123 > 115 characters)

File: corehq/apps/userreports/models.py [hunk 2]
```diff
@@ -345,6 +346,9 @@ def _check_for_duplicates(supposedly_unique_list, error_msg):
         self.charts
         self.sort_order
 
+        for agg_col in self.aggregation_columns:
```
Comment (nickpell) [inline corehq/apps/userreports/models.py]: not entirely sure if ```aggregation_columns``` get passed to the SQL query, but validated here anyway

**Comment by nickpell** [comment]
also not sure if https://github.com/dimagi/commcare-hq/pull/8722/files#diff-3af365e79b93f051d416585bcba61b7eR265 needs to be validated, or column fields for charts.  It's not clear to me out of these, which gets passed into a SQL and which doesn't

**Comment by nickpell** [comment]
convinced myself that report column ids need to be validated too, since they appear in the SQL query.  Still not sure about fields belonging to charts.

**Comment by nickpell** [comment]
Looks like charts fields do not need security validation, since they are used asynchronously by the client and never are inserted into a SQL query.

**Comment by NoahCarnahan** [comment]
Re: https://github.com/dimagi/commcare-hq/pull/8722#issuecomment-149323184
The link looks broken to me (doesn't highlight a specific line).

File: corehq/apps/userreports/models.py (hunk 2 above)
Comment (czue) [inline corehq/apps/userreports/models.py]: they should land in the `group by` so this seems legit

File: corehq/apps/userreports/util.py [hunk 3]
```diff
@@ -22,3 +27,11 @@ def localize(value, lang):
 
 def default_language():
     return "en"
+
+
+def validate_sql_column_name(s):
+    if not isinstance(s, unicode):
+        s = unicode(s, "utf-8")
+    # http://stackoverflow.com/questions/954884/what-special-characters-are-allowed-in-t-sql-column-name
```
Comment (czue) [inline corehq/apps/userreports/util.py]: noticed this question is specific to microsoft SQL. did you find any resources for postgres?

**Comment by czue** [comment]
this looks good to me - though in my head I was hoping we could get around this whole situation by using properly escaped values in the query at a lower level (e.g. in sqlalchemy). did you explore that route at all? don't have any reason to believe that it's necessarily possible, but it does seem like if we could push this problem into the library rather than in some kind of janky regexes we maintain that would be better.

also - were you able to reproduce a sql injection locally? or is this is all theoretical?

**Comment by NoahCarnahan** [comment]
Agreed it would be nice to be confident that at a low level we are protected from injections. I *think* every UCR query goes through sqlagg, in which case I think that is a good place to put the guarantees. I think SqlAlchmey makes it pretty easy to protect from injections because it escapes most strings or builds queries in an object-oriented fashion.

I think the way we handle columns is safe, but I haven't confirmed this. I'm pretty sure the way we do filters is NOT safe because we pass raw SQL to the query [here](https://github.com/dimagi/sql-agg/blob/bb2f076e060ce438170831bdf0a945582324eca6/sqlagg/base.py#L102), and that SQL is built with unsafe python string interpolation [here](https://github.com/dimagi/sql-agg/blob/bb2f076e060ce438170831bdf0a945582324eca6/sqlagg/filters.py#L30).

**Comment by nickpell** [comment]
@NoahCarnahan I tested this on filter values containing ```'``` and ```"```, and I think we're actually OK because of http://stackoverflow.com/questions/25104259/how-to-properly-escape-strings-when-manually-building-sql-queries-in-sqlalchemy

see
https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/reports/sqlreport.py#L263 and
https://github.com/dimagi/sql-agg/blob/master/sqlagg/base.py#L181

**Comment by nickpell** [comment]
@czue I didn't generate a fully-functional sql injection, but given that I was able to insert ```"```` into a doubly-quoted string without escaping, I believe it could be possible.  (it would be really nice to have a regression test if I get that far)

**Comment by nickpell** [comment]
@czue I didn't get to escaping column names in sqlagg, but I agree it would offer much more secure protection.  That said I still think there's value in validating the UCR configurations themselves

File: corehq/apps/userreports/models.py (hunk 2 above)
Comment (nickpell) [inline corehq/apps/userreports/models.py]: :+1:

File: corehq/apps/userreports/reports/specs.py (hunk 1 above)
Comment (dimagimon) [inline corehq/apps/userreports/reports/specs.py]: E501 line too long (123 > 115 characters)

File: corehq/apps/userreports/util.py (hunk 3 above)
Comment (nickpell) [inline corehq/apps/userreports/util.py]: good point - updated to reflect postgres docs.

**Comment by nickpell** [comment]
Still wrapping my head around sqlagg, will try to find the right places to add column name/alias validation tomorrow.

**Comment by NoahCarnahan** [comment]
@nickpell 
RE https://github.com/dimagi/commcare-hq/pull/8722#issuecomment-149659881

Good point, looks like the filter values are covered. However, I think the filter *names* (bind params) are not, because those are user supplied, and inserted directly into the string. In [the SO question](http://stackoverflow.com/a/25107658/1739725) you linked the filter name (`:name` in this case), was not user supplied.

**Comment by czue** [comment]
+1 to what @NoahCarnahan said. in my head [these](https://github.com/dimagi/sql-agg/blob/master/sqlagg/filters.py#L30) [places](https://github.com/dimagi/sql-agg/blob/master/sqlagg/filters.py#L40) were the biggest hole in our current code. @snopoke do you have any thoughts on any way to avoid this?

**Comment by snopoke** [comment]
We could switch to using SQLAlchemy expression functions and bindparams: http://docs.sqlalchemy.org/en/latest/core/sqlelement.html

We can also verify that column names exist in the tables. Probably somewhere around here: https://github.com/dimagi/sql-agg/blob/master/sqlagg/base.py#L102-102

**Comment by snopoke** [comment]
Looking at this PR again I'm not against this approach but it also seems like it would be easier to do 'real' validation of column names in sql-agg as mentioned above.

**Comment by nickpell** [comment]
Since this is being taken care of in sqlagg, I don't feel like it's worth complicating UCR code

**Pull request closed without merging**